### PR TITLE
Swift 4

### PR DESCRIPTION
**Pull request by mdiep**
This gets `master` building with the Swift 4 snapshots.

Changes fall into 2 categories:

- Type inference failures
- Removal of implicit tuple destructuring from SE-0110

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -55,6 +55,15 @@ jobs:
         submodules: false
       env:
         - JOB=SWIFTPM_DARWIN
+    - before_install:
+        - eval "$(curl -sL https://gist.githubusercontent.com/kylef/5c0475ff02b7c7671d2a/raw/9f442512a46d7a2af7b850d65a7e9bd31edfb09b/swiftenv-install.sh)"
+      script:
+        - swift build
```
Comment (mdiep) [inline .travis.yml]: Nimble doesn't build with Swift 4, so this only builds for now. 😞

**Comment by andersio** [comment]
Can you add a job for the Swift 3 compatibility mode?

**Comment by andersio** [comment]
(Not related to the PR) The implicit tuple destructing makes me wonder if we need to introduce sth like `combineLatestMap` or `zipMap`. 🤦‍♂️

**Comment by mdiep** [comment]
> Could you add a job for the Swift 3 compatibility mode?

I'm not sure how to do that, so I'd love to leave it for a separate PR.

**Pull request merged**